src/check/constrain/unify/ty.rs

```diff
@@ -160,14 +160,26 @@ pub fn unify_type(
             unify_link(constraints, ctx, total + 1)
         }
 
-        (l_exp, r_exp) => {
-            if l_exp.is_none() && r_exp.is_none() {
-                unify_link(constraints, ctx, total)
-            } else {
-                let msg = format!("Unifying type: Expected a '{}', was a '{}'", l_exp, r_exp);
-                Err(vec![TypeErr::new(&left.pos, &msg)])
+        (l_exp, r_exp) => match (l_exp, r_exp) {
+            (Collection { ty }, Type { name }) | (Type { name }, Collection { ty }) => {
+                if let Some(col_ty) = name.collection_type(ctx)? {
```

Comment:
If we know how many placeholders types there are, then we can just give a placeholder Type as argument (`@0`, `@1`, etc.).
So environment needs to give as output how many placeholder types there are for the current set of constraints to the unification stage.
Then we can generate a placeholder type here, pass this as argument, and let the unification stage do its thing.